src/Compilers/CSharp/Portable/Lowering/StateMachineRewriter/StateMachineRewriter.cs

```diff
@@ -152,7 +152,7 @@ protected BoundStatement Rewrite()
                     Debug.Assert(local.RefKind == RefKind.None);
                     StateMachineFieldSymbol field = null;
 
-                    if (!local.SynthesizedKind.IsSlotReusable(F.Compilation.Options.OptimizationLevel))
+                    if (ShouldPreallocateNonReusableProxy(local))
```

Comment:
In what sense is this "preallocation"?

Reply:
For some locals (and for all in debug) we create/allocate backing fields unconditionally before the actual rewrite started. And those fields are all marked as not reusable.  
The logic for selecting such locals got more complex with this change so I've moved it to a helper.